### DISCUSSION TITLE
8% HTML without a single html file in the repo

### DISCUSSION THREAD

**Discussion by shaderecker** [github-linguist/linguist#5226] [Classification] [answered]
Hi,
my repo https://github.com/shaderecker/ansible-pihole shows 8% HTML language, but there are no html files in there.
When clicking on the language tag, it shows no results (since there are no html files).

Do you have any ideas why HTML is being detected here?

This error only happens directly on GitHub.
Running linguist locally works as expected and doesn't show the wrong HTML statistics.
```
$ github-linguist . --breakdown
94.35%  YAML
5.65%   Shell

YAML:
bootstrap-pihole.yaml
inventory.yaml
keepalived.yaml
roles/bootstrap/tasks/main.yaml
roles/docker/tasks/main.yaml
roles/keepalived/tasks/main.yaml
roles/pihole/tasks/main.yaml
roles/sshd/tasks/main.yaml
roles/start_keepalived/tasks/main.yaml
roles/stop_keepalived/tasks/main.yaml
roles/sync/tasks/main.yaml
roles/updates/tasks/main.yaml
sync.yaml
update-pihole.yaml

Shell:
roles/keepalived/files/check_pihole.sh
roles/sync/templates/pihole_sync.j2
```

**Comment by lildude** [comment, marked as answer]
🤔 Looks like you might be running an old version of Linguist. This is what I see, which matches what your repo reports on GitHub:

```console
$ github-linguist ansible-pihole --breakdown
86.68%  YAML
7.37%   HTML
5.95%   Shell

YAML:
bootstrap-pihole.yaml
inventory.yaml
keepalived.yaml
roles/bootstrap/tasks/main.yaml
roles/docker/tasks/main.yaml
roles/keepalived/tasks/main.yaml
roles/pihole/tasks/main.yaml
roles/sshd/tasks/main.yaml
roles/start_keepalived/tasks/main.yaml
roles/stop_keepalived/tasks/main.yaml
roles/sync/tasks/main.yaml
roles/updates/tasks/main.yaml
sync.yaml
update-pihole.yaml

Shell:
roles/keepalived/files/check_pihole.sh
roles/sync/templates/pihole_sync.j2

HTML:
roles/keepalived/templates/keepalived.j2
$
```

`.j2` is associated with "HTML+Django" which is grouped under HTML:

https://github.com/github/linguist/blob/7e21cd427be5ab1bff72ed6e3d3834474145ea88/lib/linguist/languages.yml#L2097-L2113

Now comes the interesting question: "Why is `.j2` being detected as Shell and HTML?

Because the `roles/sync/templates/pihole_sync.j2` file has told us it's a shell script in the [very first line](https://github.com/shaderecker/ansible-pihole/blob/ebc726140b61ea120a9b1ffe2e34105f4d01b855/roles/sync/templates/pihole_sync.j2#L1) and this takes precedence over the extension.

**Reply by Alhadis** [reply]
On an unrelated side-note, [your shell-script](https://github.com/shaderecker/ansible-pihole/blob/ebc726140b61ea120a9b1ffe2e34105f4d01b855/roles/sync/templates/pihole_sync.j2#L1) should begin with

~~~sh
#!/usr/bin/env bash
~~~

instead of

~~~sh
#!/bin/bash
~~~

if you need your script to be specifically run by Bash, as opposed to `/bin/sh`, which generally maps to the system's default shell. The reason is that not every Unix-like system has `/bin/bash` installed; e.g., many BSD derivatives use `ksh(1)` or `csh(1)`, and a future release of macOS will remove the system-installed Bash once Apple have finished migrating users to Zsh.

`</diatribe>`

**Comment by shaderecker** [comment]
Thanks for the explanation @lildude 
And I was indeed running an older version (7.9.0) from Ubuntu 20.10 sources.

One question: Why does it then display no results when I click on the HTML lang tag (https://github.com/shaderecker/ansible-pihole/search?l=html)?
The `.j2` file is categorized in https://github.com/shaderecker/ansible-pihole/search?l=shell (due to the shebang)
Wouldn't it be better if it appeared in both categories then? It really confused me that there are no results for HTML.

And I also have another `.j2` file (`roles/keepalived/templates/keepalived.j2`) that I would then expect to appear in the HTML search.

And @Alhadis thanks for the hint about the shebang, I will keep it in mind.

**Reply by lildude** [reply]
> One question: Why does it then display no results when I click on the HTML lang tag (https://github.com/shaderecker/ansible-pihole/search?l=html)?
> ...
> It really confused me that there are no results for HTML.
> ...
> And I also have another .j2 file (roles/keepalived/templates/keepalived.j2) that I would then expect to appear in the HTML search.

That's because Search doesn't know about the language grouping. Linguist has said this is "HTML" (that's where the `html` comes from in the query) so that's what Search returns.

> The `.j2` file is categorized in https://github.com/shaderecker/ansible-pihole/search?l=shell (due to the shebang)
> Wouldn't it be better if it appeared in both categories then?

No. You've explicitly told Linguist that this is Shell and not HTML (or anything else) by using the shebang. Linguist will not go against this choice. Linguist can also only associate a file with one language at a time.

**Reply by shaderecker** [reply]
> That's because Search doesn't know about the language grouping. Linguist has said this is "HTML" (that's where the html comes from in the query) so that's what Search returns.

But the html search doesn't return any results. That part I don't get. 🙂

**Reply by lildude** [reply]
Whoops. Ignore what was here before. It's because the actual language is "HTML+Django" and not HTML, but Linguist is returning the group, not the language, so search is using that.

**Reply by shaderecker** [reply]
Aha ok :D Now I understand the behaviour on GitHub.
Thank you for your answers! 😃
